pkg/server/datastore/sqlstore/sqlstore.go

```diff
@@ -1052,6 +1059,60 @@ func pruneBundle(tx *gorm.DB, trustDomainID string, expiry time.Time, log logrus
 	return changed, nil
 }
 
+func taintX509CA(tx *gorm.DB, trustDomainID string, taintedPublicKey crypto.PublicKey) error {
+	model := &Bundle{}
+	if err := tx.Find(model, "trust_domain = ?", trustDomainID).Error; err != nil {
+		return sqlError.Wrap(err)
+	}
+
+	bundle, err := modelToBundle(model)
+	if err != nil {
+		return status.Errorf(codes.Internal, "failed to unmarshal bundle: %v", err)
+	}
+
+	var found bool
+	for _, ca := range bundle.RootCas {
+		rootCACert, err := x509.ParseCertificate(ca.DerBytes)
+		if err != nil {
+			return status.Errorf(codes.Internal, "failed to parse root CA: %v", err)
+		}
+
+		ok, err := cryptoutil.PublicKeyEqual(rootCACert.PublicKey, taintedPublicKey)
+		if err != nil {
+			return status.Errorf(codes.Internal, "failed to compare public key: %v", err)
+		}
+		if ok {
+			if ca.TaintedKey {
+				return status.Errorf(codes.Internal, "root CA is already tainted")
+			}
+
+			// no breaking here to be sure that we taint all
+			// bundles associated with given public key,
+			// but we must contain a single bundle signed
+			// by this key
```

Comment:
I'm not sure if I follow :)
I _think_ this could be rephrased to:

```suggestion
			// Do not break the loop here so we can be sure
			// that we taint all bundles associated with the
			// given public key.
			// We also check if there is at least one bundle
			// signed by the provided key. If not, the function
			// returns codes.NotFound status error.
```